tests/cephfs/snapshot_clone/clone_states.py

```diff
@@ -0,0 +1,129 @@
+        """
+    try:
+        fs_util = FsUtils(ceph_cluster)
+        config = kw.get("config")
+        clients = ceph_cluster.get_ceph_objects("client")
+        build = config.get("build", config.get("rhbuild"))
+        fs_util.prepare_clients(clients, build)
+        fs_util.auth_list(clients)
+        log.info("checking Pre-requisites")
+        if len(clients) < 1:
+            log.info(
+                f"This test requires minimum 1 client nodes.This has only {len(clients)} clients"
+            )
+            return 1
+        default_fs = "cephfs"
+        mounting_dir = "".join(
+            random.choice(string.ascii_lowercase + string.digits)
+            for _ in list(range(10))
+        )
+        client1 = clients[0]
+        fs_details = fs_util.get_fs_info(client1)
+        if not fs_details:
+            fs_util.create_fs(client1, "cephfs")
+        subvolumegroup_list = [
+            {"vol_name": default_fs, "group_name": "subvolgroup_clone_states_1"}
+        ]
+        for subvolumegroup in subvolumegroup_list:
+            fs_util.create_subvolumegroup(client1, **subvolumegroup)
+        subvolume = {
+            "vol_name": default_fs,
+            "subvol_name": "subvol_clone_states",
+            "group_name": "subvolgroup_clone_states_1",
+            "size": "5368706371",
+        }
+        fs_util.create_subvolume(client1, **subvolume)
+        log.info("Get the path of sub volume")
+        subvol_path, rc = client1.exec_command(
+            sudo=True,
+            cmd=f"ceph fs subvolume getpath {default_fs} subvol_clone_states subvolgroup_clone_states_1",
+        )
+        kernel_mounting_dir_1 = f"/mnt/cephfs_kernel{mounting_dir}_1/"
+        mon_node_ips = fs_util.get_mon_node_ips()
+        fs_util.kernel_mount(
+            [clients[0]],
+            kernel_mounting_dir_1,
+            ",".join(mon_node_ips),
+            sub_dir=f"{subvol_path.read().decode().strip()}",
+        )
+        client1.exec_command(
+            sudo=True,
+            cmd=f"python3 /home/cephuser/smallfile/smallfile_cli.py --operation create --threads 10 --file-size 400 "
+                f"--files 100 --files-per-dir 10 --dirs-per-dir 2 --top "
+                f"{kernel_mounting_dir_1}",
+            long_running=True,
+        )
+        snapshot = {
+            "vol_name": default_fs,
+            "subvol_name": "subvol_clone_states",
+            "snap_name": "snap_1",
+            "group_name": "subvolgroup_clone_states_1",
+        }
+        fs_util.create_snapshot(client1, **snapshot)
+        clone_status_1 = {
+            "vol_name": default_fs,
+            "subvol_name": "subvol_clone_states",
+            "snap_name": "snap_1",
+            "target_subvol_name": "clone_status_1",
+            "group_name": "subvolgroup_clone_states_1",
+        }
+        fs_util.create_clone(client1, **clone_status_1)
+        fs_util.validate_clone_state(client1, clone_status_1)
```

Comment:
Could we add separate scenarios for testing "pending" & "in-progress" states as well. As by-default, "validate_clone_state" module is testing for "complete" state only.

Reply:
Hi Yogesh,
Clone Operation is moving to **in-progess.** as soon as it is triggered.
so we are maintaining a list of valid states and verifying it.
Maybe I will try to collect all the states till it reaches a complete state. and verify the list with **valid_states**

Reply:
By storing enough data in subvolume to take some time for cloning & setting "max_concurrent_clones" lesser  & triggering cloning operation for more than "max_concurrent_clones" at same time, you can test "pending" state

Reply:
>Maybe I will try to collect all the states till it reaches a complete state. and verify the list with valid_states

Sounds good.